internal/sci/aws/server.go

```diff
@@ -0,0 +1,218 @@
+// Package sciaws provides an AWS implementation of the Substratus Cloud Interface (SCI)
+package aws
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	awsSdk "github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/eks"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/substratusai/substratus/internal/sci"
+)
+
+type Server struct {
+	sci.UnimplementedControllerServer
+	OIDCProviderURL string
+	OIDCProviderARN string
+	Clients
+}
+
+type Clients struct {
+	S3Client  *s3.S3
+	IamClient *iam.IAM
+}
+
+func (s *Server) GetObjectMd5(ctx context.Context, req *sci.GetObjectMd5Request) (*sci.GetObjectMd5Response, error) {
+	// ensure the object is accessible
+	bucketName, objectName := req.GetBucketName(), req.GetObjectName()
+	input := &s3.HeadObjectInput{
+		Bucket: awsSdk.String(bucketName),
+		Key:    awsSdk.String(objectName),
+	}
+	headResult, err := s.Clients.S3Client.HeadObject(input)
+	if err != nil {
+		return nil, err
+	}
+
+	// NOTE: AWS returns an MD5 checksum as an ETag except for multi-part uploads where it's an MD5 with a dash suffix.
+	if headResult.ETag == nil {
+		return nil, fmt.Errorf("object does not exist: %s", s3.ErrCodeNoSuchKey)
+	}
+
+	md5 := *headResult.ETag
+
+	return &sci.GetObjectMd5Response{
+		Md5Checksum: md5,
+	}, nil
+}
+
+func (s *Server) CreateSignedURL(ctx context.Context, req *sci.CreateSignedURLRequest) (*sci.CreateSignedURLResponse, error) {
+	bucketName, objectName, checksum := req.GetBucketName(),
+		req.GetObjectName(),
+		req.GetMd5Checksum()
+
+	// ensure the object is accessible
+	input := &s3.HeadObjectInput{
+		Bucket: awsSdk.String(bucketName),
+		Key:    awsSdk.String(objectName),
+	}
+	_, err := s.Clients.S3Client.HeadObject(input)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok {
+			if awsErr.Code() != s3.ErrCodeNoSuchKey {
+				return nil, fmt.Errorf("failed to head object: %w", err)
+			}
+		} else {
+			// It's an error of a different type, not an awserr.Error
+			return nil, fmt.Errorf("failed to head object: %w", err)
+		}
+	}
```

Comment:
Like in the gcs sci, I added this to catch errors related to accessing this key path.

Some S3 errors we might catch:

Permission issues (maybe the provided credentials don't have the required permission).
Network issues (maybe there's an issue connecting to the S3 service).
Bucket issues (unlikely but potentially the bucket doesn't exist).
AWS SDK or client-related issues.

Non-AWS Errors:
Errors not related to the AWS SDK. Potentially lower-level issues.